pkg/odo/cli/component/create.go

```diff
@@ -905,8 +907,8 @@ func (co *CreateOptions) devfileRun() (err error) {
 }
 
 // Run has the logic to perform the required actions as part of command
-func (co *CreateOptions) Run() (err error) {
-
+func (co *CreateOptions) Run(cmd *cobra.Command) (err error) {
+	segment.SetComponentType(cmd.Context(), co.devfileMetadata.componentType)
```

Comment:
why change the signature of run to get access to the `context.Context`? I would suggest maybe setting it as part of the `CreateOptions` struct
```
type CreateOptions struct {
cmdContext context.Context
}
```

then setting it inside the `Complete` function which gets the `cmd` as an argument.

Reply:
Yup, there is cmd in Complete(), but I don't want to upload information about a component type that might have been misspelled or some other error, I want to wait until it is ready for Run(). But I'll look into this https://github.com/openshift/odo/pull/4662#discussion_r626018142. See if it's possible.

Reply:
You can validate the name in complete for this case only. We have made exceptions when it’s necessary

Reply:
Yeah so I would suggest just set the context in Complete,
Then validate the type in Validate but send it in Run. 
as you would have the context already in the struct you don’t need to change the interface

Reply:
@girishramnani I tried what you suggested here but it doesn't make sense to me to have one more context added to co, plus,  you suggested we set `co.cmdContext` in `Complete()` and then send it in `Run()` but that is not possible as far as I can tell because the value is sent from `runnable.go::startTelemetry` which calls `segment.GetContextProperties(context.Context)`, we need to pass the same context in which we set a value before, in our case, `co.cmdContext`. I do not think there is a way to pass that value in `startTelemetry`.

Also, having `cmd` in `Run()` will help when we want to pass other values from other runnables.

Reply:
> why change the signature of run to get access to the `context.Context`? I would suggest maybe setting it as part of the `CreateOptions` struct
> 
> ```
> type CreateOptions struct {
> cmdContext context.Context
> }
> ```

No, please don't do this, as this is not the recommended way to pass contexts: 

"Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:" from https://golang.org/pkg/context/